milli/src/update/facets.rs

```diff
@@ -151,178 +260,332 @@ fn compute_facet_number_levels<'t>(
+* `Bound`: part of the range in the levels structure. For example, for numbers, the `Bound` is `f64`
+because each chunk in a level contains a range such as (1.2 ..= 4.5).
+
+## Arguments
+* `rtxn` : LMDB read transaction
+* `db`: a database which already contains a `level 0`
+* `compression_type`/`compression_level`: parameters used to create the `grenad::Writer` that
+will contain the new levels
+* `level` : the height of the level to create, or `0` to read elements from level 0.
+* `level_0_start` : a key in the database that points to the beginning of its level 0
+* `level_0_range` : equivalent to `level_0_start..`
+* `level_0_size` : the number of elements in level 0
+* `level_group_size` : the number of elements from the level below that are represented by a
+* single element of the new level
+* `computed_group_bitmap` : a callback that is called whenever at most `level_group_size` elements
+from the level below were read/created. Its arguments are:
+    0. the list of bitmaps from each read/created element of the level below
+    1. the start bound corresponding to the first element
+    2. the end bound corresponding to the last element
+* `bound_from_db_key` : finds the `Bound` from a key in the database
+* `bitmap_from_db_value` : finds the `RoaringBitmap` from a value in the database
+* `write_entry` : writes an element of a level into the writer. The arguments are:
+    0. the writer
+    1. the height of the level
+    2. the start bound
+    3. the end bound
+    4. the docids of all elements between the start and end bound
+
+## Return
+A vector of grenad::Reader. The reader at index `i` corresponds to the elements of level `i + 1`
+that must be inserted into the database.
+*/
+fn recursive_compute_levels<'t, KeyCodec, ValueCodec, Bound>(
+    rtxn: &'t heed::RoTxn,
+    db: heed::Database<KeyCodec, ValueCodec>,
+    compression_type: CompressionType,
+    compression_level: Option<u32>,
+    level: u8,
+    level_0_start: <KeyCodec as BytesDecode<'t>>::DItem,
+    level_0_range: &'t RangeFrom<<KeyCodec as BytesDecode<'t>>::DItem>,
+    level_0_size: usize,
+    level_group_size: NonZeroUsize,
+    computed_group_bitmap: &mut dyn FnMut(&[RoaringBitmap], Bound, Bound) -> Result<()>,
+    bound_from_db_key: &dyn for<'a> Fn(usize, &'a <KeyCodec as BytesDecode<'t>>::DItem) -> Bound,
+    bitmap_from_db_value: &dyn Fn(<ValueCodec as BytesDecode<'t>>::DItem) -> RoaringBitmap,
+    write_entry: &dyn Fn(&mut Writer<File>, NonZeroU8, Bound, Bound, RoaringBitmap) -> Result<()>,
+) -> Result<Vec<Reader<File>>>
+where
+    KeyCodec: for<'a> BytesEncode<'a>
+        + for<'a> BytesDecode<'a, DItem = <KeyCodec as BytesEncode<'a>>::EItem>,
+    for<'a> <KeyCodec as BytesEncode<'a>>::EItem: Sized,
+    ValueCodec: for<'a> BytesEncode<'a>
+        + for<'a> BytesDecode<'a, DItem = <ValueCodec as BytesEncode<'a>>::EItem>,
+    for<'a> <ValueCodec as BytesEncode<'a>>::EItem: Sized,
+    Bound: Copy,
```

Comment:
This is one of the reasons why I would like to rethink the `ByteEncode/ByteDecode` trait to be simpler, if possible, and be directly implemented on the serialized types instead of serializer types ([kind of like sanakirja](https://docs.rs/sanakirja/latest/sanakirja/trait.Storable.html)).

Reply:
I would also like to rethink the `BytesEncode/BytesDecode` traits :-)